thegame/engine/engine.py

```diff
@@ -72,33 +90,102 @@ def _main_loop(self):
+                    self.context.active_screen
+                )
+                logging.debug(
+                    f"Got the following fov_tile_sheet: {pprint.pformat(fov_tile_sheet, indent=4)}"
+                )
+
+                # Currently, this is done with (4 * camera_width * camera_height) loops.
+                # the only way to reduce this is to not loop over rows full of Nones, but
+                # this check isn't worth the gain.
+                # Updates each sprites' x and y position.
+                for layer in fov_tile_sheet:
+                    for row_index, row in enumerate(layer):
+                        for cell_index, cell in enumerate(row):
+                            if cell is not None:
+                                logging.debug(
+                                    f"{cell_index}: Setting the position of a {type(cell).__name__} on the screen."
+                                )
+                                cell.set_sprite_position(
+                                    cell_index * self.context.base_sprite_width,
+                                    row_index * self.context.base_sprite_height,
+                                )
+                                onscreen_sprites.append(cell.get_sprite())
+
+                for sprite in onscreen_sprites:
+                    game_sprites.add(sprite)
+
+            game_sprites.draw(self.display)
+
+            # TODO: In conjunction with the above todo, a more efficient way of drawing should be found.
+            pygame.display.flip()
+
+            # Run at 60fps
+            game_clock.tick(60)
+
+    def _load_map_sprites(self):
+
+        game_object_list_set = set()
+
+        # TODO: This is really ugly, and probably not optimal
+        #       This should be re-looked at to be improved.
+        for game_map in self.context.maps.values():
+            for layer in game_map.tile_sheets:
+                for row in layer:
+                    for game_object in row:
+                        if game_object is not None:
+                            game_object_list_set.add(game_object.sprite_location)
+
+        for game_object_location in game_object_list_set:
+            image = pygame.image.load(game_object_location).convert_alpha()
+
+            self.context.object_images[game_object_location] = image
+
+    def _load_menu_sprites(self):
+        for menu in self.context.menus.values():
+            sprite = pygame.sprite.Sprite()
+            sprite.image = pygame.image.load(menu.menu_image_location).convert_alpha()
+            sprite.rect = sprite.image.get_rect()
+            menu.menu_image = sprite
+
+    def _schedule_events(self, events, number_of_events):
+        # Each event should be independent of the other,
+        # thus we can process each one as if the others
+        # didn't happen.
+        event_pool = ThreadPool(processes=number_of_events)
+
+        exception_check = []
```

Comment:
Might want to add a comment for the weird exception rules that happen here.

Reply:
Will do.